### DISCUSSION TITLE
Software Simulator

### DISCUSSION THREAD

**Discussion by gramster** [SynthstromAudible/DelugeFirmware#628] [Ideas]
This is no small task, but one where I would be more interested to contribute, rather than modding my device. Questions have come up about governance, standards for PR acceptance, and stability. The usual way to mitigate some of these concerns is through extensive automated testing. That is hard to do with a hardware device unless you have a simulator. A simulator in this case might have other positive spin-offs (think, Deluge VST, for example).

In order to get there, hardware interfaces need to be reasonable well abstracted; you don't want to be poking random memory addresses all over the place, for example. I'm still new to Deluge, have little familiarity with the hardware or source code, so don't know what state the current software is in in this respect; would love some pointers from those who do know the source and low-level hardware better.

**Comment by PaulFreund** [comment]
Hey there! :) Executing the Deluge application in a simulated/mocked environment would be awesome but very far along the way. The codebase is missing some critical abstraction layers and there is still a lot of work in stabilizing. 

My take is that it would make most sense to start an application architecture project first that works in multiple environments and then start to port features from the Deluge application to it. Benefits would be platform portability, testability, being able to split out certain things (e.g. have the Synth as a VST, the whole UX as an Android App etc.)

Still a lot of daydreaming, feel free to join the Discord to join in :D
